distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+                         debug=False):
+    """ Detect outlier genes algorithm described in section "Detecting
+    `Outlier' Genes" of the Wei. X et al. paper. The full distance matrix is
+    represented in the format:
+
+    full_distance_matrix[#genes][#species][#species] =
+    [[[0_0, 0_1, 0_2, .., 0_n]
+      [1_0, 1_1, 1_2, .., 1_n]
+      ..
+      [n_0, n_1, n_2, .., n_n]]
+
+     [[0_0, 0_1, 0_2, .., 0_n]
+      [1_0, 1_1, 1_2, .., 1_n]
+      ..
+      [n_0, n_1, n_2, .., n_n]]
+
+      ..
+     [[0_0, 0_1, 0_2, .., 0_n]
+      [1_0, 1_1, 1_2, .., 1_n]
+      ..
+      [n_0, n_1, n_2, .., n_n]]]
+
+    The mean and standard deviation are computed for each species pair
+    including all genes.
+
+    Parameters
+    ----------
+    species_set: list
+        list of bitvectors representing species clusters to use in detecting
+        outlier genes
+    gene_bitvector_map: list
+        list containing the binary indicator vector for each query gene
+    full_distance_matrix: dictionary
+        complete distance matrix for pairwise alignments between all species
+        for every gene
+    stdev_offset: integer
+        the number of standard deviations a gene's normalized distance is from
+        the mean to identify it as an outlier for a species pair
+    outlier_hgt: float
+        the fraction (value between (0,1]) of normalized pairwise distances
+        over all species-pair vectors belonging to the same gene that are
+        z-score standard deviations from the mean
+    num_species: integer
+        number of species in the reference database
+    total_genes: integer
+        total number of genes in the query genome with at least
+        min_num_homologs (determined by BLAST search)
+    debug: boolean
+        if True, run function in debug mode
+
+    Returns
+    -------
+    outlier_genes: set
+        set of atypical genes
+    """
+    numpy.around(full_distance_matrix, decimals=5, out=full_distance_matrix)
+    outlier_flag_matrix = numpy.zeros(
+        shape=(total_genes, num_species, num_species), dtype=bool)
+    distance_vector = numpy.zeros(total_genes)
+    if debug:
+        sys.stdout.write("[DEBUG] species_species\t")
+        for k in range(total_genes):
+            sys.stdout.write("gene # %s".ljust(12) % k)
+        sys.stdout.write("[low_bound, up_bound]\n")
+    for i in range(num_species):
+        for j in range(num_species):
+            if i != j:
+                for k in range(total_genes):
+                    distance_vector[k] = full_distance_matrix[k][i][j]
+                mean = numpy.nanmean(distance_vector)
+                stdev = numpy.nanstd(distance_vector)
+                low_bound = round(mean - stdev_offset*stdev, 5)
+                up_bound = round(mean + stdev_offset*stdev, 5)
+                if debug:
+                    sys.stdout.write("[DEBUG] %s_%s\t".ljust(20) % (i, j))
+                for k, distance in enumerate(distance_vector):
```

Comment:
My guess is that at the end of this for you could add the next two fors (704-719) and avoid some loops